flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerStartupTest.java

```diff
@@ -0,0 +1,210 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.CoreOptions;
+import org.apache.flink.configuration.IllegalConfigurationException;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.runtime.blob.BlobCacheService;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.heartbeat.HeartbeatServices;
+import org.apache.flink.runtime.highavailability.HighAvailabilityServices;
+import org.apache.flink.runtime.metrics.NoOpMetricRegistry;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.util.IOUtils;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.commons.io.FileUtils;
+import org.junit.Assume;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeNoException;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests that check how the {@link TaskManagerRunner} behaves when encountering startup problems.
+ */
+public class TaskManagerRunnerStartupTest extends TestLogger {
+
+	private static final String LOCAL_HOST = "localhost";
+
+	@Rule
+	public TemporaryFolder tempFolder = new TemporaryFolder();
+
+	private RpcService rpcService = createRpcService();
+
+	/**
+	 * Tests that the TaskManagerRunner startup fails synchronously when the I/O
+	 * directories are not writable.
+	 */
+	@Test
+	public void testIODirectoryNotWritable() throws Exception {
+		File nonWritable = tempFolder.newFolder();
+		Assume.assumeTrue("Cannot create non-writable temporary file directory. Skipping test.",
+			nonWritable.setWritable(false, false));
+
+		try {
+			Configuration cfg = new Configuration();
+			cfg.setString(CoreOptions.TMP_DIRS, nonWritable.getAbsolutePath());
+
+			try {
+				TaskManagerRunner.startTaskManager(
```

Comment:
this call is identical in all tests and could be moved into a static factory method